fix: Unify current installation

kanishk16 opened the pull request:
## Description
**tl;dr**: Unifies the installation process through `extras_require` argument in `setup.py`

As discussed in #996 and in light of discussion in #795 especially @kousu's [comment](https://github.com/ivadomed/ivadomed/issues/795#issuecomment-879952457) making torch non-mandatory resulted in a few blockers as well as overgrown installation instructions. Thus, increasing the priority to streamline the installation process along with mandating torch installation keeping in mind the user's freedom to choose a backend for torch either through CPU or GPU. 

This PR:
1. Reverts #757 by eliminating the need for cascading `requirements.txt` files and making torch mandatory through `extras_require`
2. Resolves #861 as users can now install ivadomed through pip via `pip install ivadomed[cpu]` or `pip install ivadomed[gpu]` which not only installs minimal requirements but also torch depending upon the backend (still pinned to CUDA-11). Moreover, `pip install ivadomed[dev_gpu]` installs minimal requirements + torch on GPU (CUDA-11) + dev requirements 
3. Consequently, a limitation of this plausible solution is `pip install ivadomed` wouldn't install anything requiring users to explicitly specify the backend to install ivadomed. However, it makes easier for downstream packages to specify ivadomed as a dependency `ivadomed[cpu]` or `ivadomed[gpu]`. 
4. ~#1057 as proposed in the [comment ](https://github.com/ivadomed/ivadomed/issues/1057#issuecomment-1031707720). As the proposed solution is much more general and handles concerns to Compute Canada, the idea of `pip install ivadomed[inference]` is dropped. `pip install ivadomed[inference]` installs minimal requirements + torch and torchvision on cpu this can be further tuned to install packages that would be required for inference only, related to #795.~
Update: Moved to Singularity as a fix since ERROR(s) still persist
5. Resolves #996

HOW?
This PR introduces two functions to `setup.py`:
1. `get_whl_info()` - returns tags required to handpick torch +torchvision wheels
2. `get_torch_whls(backend)` - handpicks torch + torchvision wheels based upon the backend

#### TODO:
- [x] Update Docs: Open a separate PR to incorporate the updated procedure for installation and all the minor readability issues

To test the PR, just try any of the following commands to install ivadomed (preferably in a new virtual environment):
- `pip install .[cpu]` - installs minimal requirements + torch on CPU
- `pip install .[gpu]` - installs minimal requirements + torch on GPU (CUDA-11)
- `pip install .[dev_cpu]` - installs minimal requirements + torch on CPU + dev requirements (development, doc & test related dependencies)
- `pip install .[dev_gpu]` - installs minimal requirements + torch on GPU (CUDA-11) + dev requirements (development, doc & test related dependencies)


## Checklist

#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Linked issues

kanishk16 committed:
fix: unify current installation

kanishk16 committed:
refactor: remove distutils as deprecated

kanishk16 committed:
move to extras_require as a fix

kanishk16 committed:
add deleted requirements.txt

kanishk16 commented:
cc @dyt811 It would be great if you could test this fix on macOS and would love to have your opinions on the plausible solution.

dyt811 commented:
Great job. That should extremely simplify the process with installation. 

I agree testing on mac (which doesn't have GPU support) which relies on its `+cpu` tag is important if we plan to support that. The solution is much more elegant and sustainable over time. I only have BigSur on a 3rd party laptop so I will try to find a time to test it out. However, the unit tests by GitHub action also include Big Sur and 10.7 unit tests as well so you can leverage those too. The bigger issue potentially is with we can't test it until we finally make a release. A test release to `PyPi` may be necessary. 

You might want to also update ReadTheDoc as they rely on a requirement text file to build Sphinx doc (due to the extra requirement sphinx needs to install etc).  We may also need to update instructions there or in the YML?

https://github.com/ivadomed/ivadomed/blob/yd/multi_session_support_reforged/.readthedocs.yml

I just checked RTD, it seems most settings are going to be at other places than its interface but let me know if you feel certain settings can only be changed there to help pass unit tests. 
![image](https://user-images.githubusercontent.com/6720068/146712262-8e86b5bb-4464-4c00-ae0f-c00eb6609ca4.png)

Speaking of unit tests, you might also need to update the GitHub YAML here https://github.com/ivadomed/ivadomed/blob/master/.github/workflows/run_tests.yml 

This should eventually make the installation documentation a breeze to setup. Great stuff!

kanishk16 committed:
update dependencies following data augmentation and #1048

kanishk16 committed:
refactor and update CI/CD

kanishk16 committed:
merge master after resolving conflicts

kanishk16 committed:
refactor & move flake8 to `requirements.txt`

kanishk16 committed:
remove `requirements_common.txt` & `requirements_dev.txt`

kanishk16 committed:
refactor to conform with current installation style

kanishk16 committed:
fix typo

kanishk16 committed:
refactor to support torch whls for macos

kanishk16 committed:
initiate fix to support macos

kanishk16 committed:
refactor fallback

kanishk16 committed:
revert refactor

kanishk16 committed:
fix to support onnxruntime on cc

kanishk16 committed:
refactor for modularity

kanishk16 committed:
update `pre-commit.yaml`

kanishk16 committed:
update `.readthedocs.yml`

kanishk16 committed:
fix for `.readthedocs.yml`

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+            install torch==1.8.0+cpu, torchvision==0.9.0+cpu or \
+            torch==1.8.1+cu111, torchvision==0.9.1+cu111 from \
+            https://pytorch.org/get-started/previous-versions/")
+    elif sysinfo.system == 'Windows' and sysinfo.machine == 'AMD64':
+        platform = 'win_amd64'
+        return python_tag, abi_tag, platform
```
Maybe we should raise on x86? I don't have a 32bit windows to check this though :(. We can just say maybe 32bit not supported? Don't need to get into the details.

kanishk16 replied:
Lol even I didn't know about it but "PyTorch doesn’t work on 32-bit system. Please use Windows and Python 64-bit version." as per [official PyTorch FAQs](https://pytorch.org/docs/stable/notes/windows.html). Since, `sysinfo = uname()` and `uname()` still returns `AMD64` on a 64-bit machine with intel processor. Moreover, I couldn't find the output of `uname()` for 32-bit Windows machine by Googling. What about when someone raises an issue regarding this, we then update?

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+        platform = 'win_amd64'
+        return python_tag, abi_tag, platform
+    elif sysinfo.system == 'Darwin':
+        # TODO: update mac conditions as we upgrade to higher torch versions
+        # with separate binaries for 10 and 11 versions   
+        # macinfo = mac_ver()
+        # if macinfo[0].startswith('10') and macinfo[2] == 'x86_64':
```
Maybe we implemented  the `if` condition you have here and warn if it is not 10/9 etc and just raise as above mentioning it is not supported?

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+            torchvision_whl = f'torchvision@https://download.pytorch.org/whl/torchvision-0.9.0{_torchvision_whl}'
+            return torch_whl, torchvision_whl
+        else:
+            whl = f'%2B{backend}-{python_tag}-{abi_tag}-{platform}.whl'
```
Nice job decoding the common patterns from torch whl! Just curious, do you know if these whl patterns are universals or just particular to Torch/Torchvision?

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+        else:
+            backend = 'cu111'
+            whl = f'%2B{backend}-{python_tag}-{abi_tag}-{platform}.whl'
+            torch_whl = f'torch@https://download.pytorch.org/whl/cu111/torch-1.8.1{whl}'
+            torchvision_whl = f'torchvision@https://download.pytorch.org/whl/cu111/torchvision-0.9.1{whl}'
```
Was wondering why these are 1.8.1 while others are 1.8.0 (similar question for torchvision). Also, do you think we should potentially parameterize this number out such that if we decide to move all platform to 1.9 or other higher number, it should handle equally in the future across all platforms?

kanishk16 replied:
Just replicated whatever versions we had currently 🙂

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+extra_requirements = {
+    'cpu': [
+        requirements,
+        'onnxruntime==1.7.0',
```
So if I am reading correctly, we are essentially now discouraging people from using `requirements.txt` and migrating primarily to `pip install` with appropriate tag correct, and if they do `pip install -r requirements.txt` it will actually miss the torch stuff as well as the `onnxruntime` as by design now, correct?

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+    'cpu': [
+        requirements,
+        'onnxruntime==1.7.0',
+        get_torch_whls(backend='cpu')
```
SLICK!

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+    ],
+    'gpu': [
+        requirements,
+        get_torch_whls(backend='gpu'),
```
SLICKKK!

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+        requirements,
+        get_torch_whls(backend='gpu'),
+    ],
+    'docs': [
+        # pin sphinx to match what RTD uses:
+        # https://github.com/readthedocs/readthedocs.org/blob/ecac31de54bbb2c100f933e86eb22b0f4389ba84/requirements/pip.txt#L16
+        'sphinx',
+        'sphinx_rtd_theme',
+        'sphinx-tabs==3.2.0',
+        'sphinx-toolbox==2.15.2',
+        'sphinx-jsonschema~=1.16',
+        'pypandoc',
```
Nice! 👍 Yeah, this is much more elegant than using extra requirement to handle RTD.

~~We may need to ensure Devs also include this in their local install. Is there a way to like... install all the extra requirements?~~

Addressed below. Perfect.

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+        'sphinx-jsonschema~=1.16',
+        'pypandoc',
+    ],
+    'tests': [
+        'pytest~=6.2',
+        'pytest-cov',
+        'pytest-ordering~=0.6',
+        'pytest-console-scripts~=1.1',
+        'coverage',
+        'coveralls',
+    ],
+    'contrib': [
+        'pre-commit>=2.10.1',
+        'flake8',    
+    ]
```
~~These are pretty much required for devs as well? So probably in the PR on documentation update, we should mention this on the installation for dev section.~~

I saw the lines below. Perfect!

dyt811 commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+    ]
+}
+
+extra_requirements['dev_cpu'] = [
+    extra_requirements['cpu'],
+    extra_requirements['docs'],
+    extra_requirements['tests'],
+    extra_requirements['contrib'],
+    ]
+
+extra_requirements['dev_gpu'] = [
+    extra_requirements['gpu'],
+    extra_requirements['docs'],
+    extra_requirements['tests'],
+    extra_requirements['contrib'],
+    ]
```
NICE! That is very elegant.

dyt811 commented on setup.py:
```diff
@@ -35,15 +161,7 @@
-        ],
-        'dev': ["pre-commit>=2.10.0"]
-    },
+    extras_require=extra_requirements,
```
MUCH cleaner.

dyt811 reviewed (commented):
Thank you for the great work. TIL much about how to improve installation `setup.py`. Very nicely done. Going to have to run a bit more GPU based manual testing and should be good to merge. 

- [X] Windows CPU Testing all green! 
- [ ] Windows GPU Testing failed with logs shown below. (Tested on a brand new CondaEnv)
   *  [GPU_TestingFailure.txt](https://github.com/ivadomed/ivadomed/files/8150827/GPU_TestingFailure.txt)</details>

- [ ] Linux GPU Testing also failed with logs shown below. (Tested on a new CondaEnv)
   * [LinuxGPUInstallationFail.txt](https://github.com/ivadomed/ivadomed/files/8151080/LinuxGPUInstallationFail.txt)

The device ordinal issue sounds like a `config.json` whic hmight needs to be updated (as it is only used in GPU instanced, it bypass CI). 

As for the onnx? I am not sure? I think did somehow `[dev_gpu]` missed that? Both systems I am usinig that as it seems to cover almost everything needed.

dyt811 reviewed (changes requested):
I think we probalby still need to adjust the Onnxruntime? Also probably good to update the GPU oridinal. Oddly it was asking 7th GPU? Beside those minor points, I think everything looks great!

jcohenadad commented:
> pip install .[cpu] - installs minimal requirements + torch on CPU
pip install .[gpu] - installs minimal requirements + torch on GPU (CUDA-11)
pip install .[dev_cpu] - installs minimal requirements + torch on CPU + dev requirements
pip install .[dev_gpu] - installs minimal requirements + torch on GPU (CUDA-11) + dev requirements

can you please clarify what 'dev' means? what is there than non-dev folks won't have?

also, as mentioned in the lab meeting, why not making cpu installed by default for everyone (since everyone has a cpu)? then, ppl _with_ a gpu could  select gpu installation

dyt811 commented:
@kanishk16 also, just FYI. On Compute Canada, the GPU installation is failing because TIFFFILE may not be available for the Compute Canada environment: 
![image](https://user-images.githubusercontent.com/6720068/157092499-d4386904-a75b-49e7-9616-a94aa85e1cf7.png)

This was done using your branch. 
![image](https://user-images.githubusercontent.com/6720068/157092651-fb2c952e-350c-4472-b9d3-8b65be268e96.png)

This confirms earlier issues with Compute Canada but may be resolved via Singularity containers (which hard bundles all its dependencies) which bypass this issue for you if we  decide to adopt that approach.

kanishk16 committed:
add feedback:
- move `onnxruntime` to `requirements.txt`
- revert back changes in `inference.py`

kousu commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+
+extra_requirements = {
+    'cpu': [
+        requirements,
```
I think this is redundant? This part is already included in `install_requires` isn't it?

kousu commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+        requirements,
+        get_torch_whls(backend='cpu')
+    ],
+    'gpu': [
+        requirements,
+        get_torch_whls(backend='gpu'),
```
I think this list needs to be flattened? So it should read more like

```suggestion
    'gpu': list(requirements) + list(get_torch_whls(backend='gpu')),
```

And if https://github.com/ivadomed/ivadomed/pull/1000/files#r866342833 is right, then actually just

```suggestion
    'gpu': get_torch_whls(backend='gpu'),
```

kousu replied:
er, and you need to cut the following `]` too of course

kousu replied:
I'm using [this as my syntax reference](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html?highlight=extras_require#optional-dependencies)

kousu commented:
Hello, thank you for your work. This will help SCT out a lot! We've been keeping an eye closely on https://github.com/pytorch/pytorch/issues/26340, which we think is the only good long term solution to the torch cpu/gpu problem. But we'll take a solid workaround in the meantime.

I warned in the meeting with @dyt811 et al. today that adding procedural code to setup.py is pretty heavily deprecated. But I need to say mea culpa. It's true that PyPA really wants people to use [setup.cfg](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html?highlight=declarative) instead of setup.py,. However, I was wrong: this code works because your procedural code here is used to **generate** declarative code:

https://github.com/ivadomed/ivadomed/blob/06b8e66e1c0701b7e4f58232e6164ea35fa466da/setup.py#L99-L107

and that works fine for end users.

So I think this is a really really excellent achievement, and thank you so much @kanishk16 for your work. I can't wait to see it integrated into SCT.

I'm still a bit surprised this works [because](https://github.com/pypa/pip/issues/5898#issuecomment-674370312)

> No, but the _end user_ needs to explicitly opt into any other index being used. This was a deliberate policy decision, to prevent malicious code on PyPI triggering download of code from other, arbitrary, locations.

i.e. PyPA feels strongly that the only safe package server is pypi.org, and anything else requires user opt-in (and thus user cognitive load, which is what we're trying to fight against I think). But here you're clearly working around that by using the `@` syntax:


https://github.com/ivadomed/ivadomed/blob/06b8e66e1c0701b7e4f58232e6164ea35fa466da/setup.py#L94-L95

kousu commented:
> I'm still a bit surprised this works [because](https://github.com/pypa/pip/issues/5898#issuecomment-674370312)
> 
> > No, but the _end user_ needs to explicitly opt into any other index being used. This was a deliberate policy decision, to prevent malicious code on PyPI triggering download of code from other, arbitrary, locations.

Ah, turns out I did an experiment on this almost 2 years ago, and while this might build a working .whl, pypi.org **does not allow you to upload such a .whl**:

https://github.com/pypa/pip/issues/5898#issuecomment-674450350

So I don't think this will work for us, not if we want SCT and ADS to be published to PyPI eventually.

kanishk16 commented:
> > I'm still a bit surprised this works [because](https://github.com/pypa/pip/issues/5898#issuecomment-674370312)
> > > No, but the _end user_ needs to explicitly opt into any other index being used. This was a deliberate policy decision, to prevent malicious code on PyPI triggering download of code from other, arbitrary, locations.
> 
> Ah, turns out I did an experiment on this almost 2 years ago, and while this might build a working .whl, pypi.org **does not allow you to upload such a .whl**:
> 
> [pypa/pip#5898 (comment)](https://github.com/pypa/pip/issues/5898#issuecomment-674450350)
> 
> They [emphasize](https://github.com/pypa/pip/issues/5898#issuecomment-674457168)
> 
> > Basically, a user should be in charge of where they download files from, it should not be something under someone else's control. Anything that takes that control away from end users is not going to come back and if the URL form of PEP 508 works on PyPI, that's a bug that needs fixed. It should not work for any type of distribution uploaded to PyPI.
> 

Thanks for the beautiful thread. I definitely missed this! I'll give it a try on a different repo altogether just to have a first-hand experience.

> So I don't think this will work in its current form for us, not if we want SCT and ADS to be published to PyPI eventually. And for that I think we need to put pressure on torch over at [pytorch/pytorch#26340](https://github.com/pytorch/pytorch/issues/26340). Or maybe we can just take their .whls and upload them ourselves to PyPI under a package named `neuropoly-torch-cpu`?

After reviewing a lot of packages that depend upon PyTorch, it appears most of them have taken a workaround that we are currently following by suggesting users to separately downloading torch. Interestingly, they have started pushing torch cpu whls on PyPI post 1.11.0 release. Perhaps we could upgrade ivadomed to v1.11.0 which might solve SCTs' problem but neither ours nor ADSs'.

kousu commented on setup.py:
```diff
@@ -1,23 +1,149 @@
+            return torch_whl, torchvision_whl
+        else:
+            whl = f'%2B{backend}-{python_tag}-{abi_tag}-{platform}.whl'
+            torch_whl = f'torch@https://download.pytorch.org/whl/cpu/torch-1.8.0{whl}'
```
would this work to get this uploadable to pypi?

```suggestion
            torch_whl = f'torch==1.8.0+cpu'
```


And then instruct users:

`pip install --find-links https://download.pytorch.org/whl/torch ivadomed`

?

kousu replied:
And sct would say:

```
pip install --find-links https://download.pytorch.org/whl/torch spinalcordtoolbox
```

kousu commented:
> Thanks for the beautiful thread. I definitely missed this! I'll give it a try on a different repo altogether just to have a first-hand experience.

Thank you :) .Sorry it had to burst your bubble though 🫧, @kanishk16.

> After reviewing a lot of packages that depend upon PyTorch, it appears most of them have taken a workaround that we are currently following by suggesting users to separately download torch.

Thanks for doing that review! That's disappointing but not surprising.

The problem for us over in SCT is that to override torch to give our Linux users a good experience ((and a lot of our users *are* on Linux, clinicians trying to improve their workflow using Ubuntu or maybe WSL)) we need to [specify the full version of torch too](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/528dfe2f0c658f7a8a494ad2cc90c37c10c29b47/requirements.txt#L38):

```
torch==1.5.0+cpu; sys_platform != "darwin"
```

So this currently means we're downgrading torch on you, because we've fallen behind you.

That's why we were hoping we could instead specify

```
torch[cpu]
```

or if not that (since it looks like Facebook Research is not down with making that happen) then

```
ivadomed[cpu]
```

and then torch would be a compartmentalized concern of ivadomed and we wouldn't have the maintenance burden in SCT, which is what this PR would have given us.

kousu commented:
I think in SCT we could maybe live with telling people to do `pip install --extra-index-url spinalcordtoolbox`, or, as we currently do, [hiding that in a requirements.txt](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/528dfe2f0c658f7a8a494ad2cc90c37c10c29b47/requirements.txt#L32-L39). It will mean that we also need to teach our downstream depencencies that they need to tell their users to do the same thing, but we're still pretty far from being able to have proper downstream dependencies anyway (https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2669).

> Interestingly, they have started pushing torch cpu whls on PyPI post [1.11.0 release](https://pypi.org/project/torch/1.11.0/) . Perhaps we could upgrade ivadomed to v1.11.0 which might solve SCTs' problem but neither ours nor ADSs'.

I'm not sure I follow this. In [1.11.0](https://pypi.org/project/torch/1.11.0/#files) I see four platforms (times four python versions, but here I'm just quoting the python3.7 part)

> * torch-1.11.0-cp37-none-macosx_10_9_x86_64.whl (129.9 MB view hashes)
> * torch-1.11.0-cp37-cp37m-win_amd64.whl (158.0 MB view hashes)
> * torch-1.11.0-cp37-cp37m-manylinux2014_aarch64.whl (51.0 MB view hashes)
> * torch-1.11.0-cp37-cp37m-manylinux1_x86_64.whl (750.6 MB view hashes)

In [1.8.0](https://pypi.org/project/torch/1.8.0/#files) I also see four platforms:

> * torch-1.8.0-cp37-none-macosx_10_9_x86_64.whl (119.5 MB view hashes)
> * torch-1.8.0-cp37-cp37m-win_amd64.whl (190.5 MB view hashes)
> * torch-1.8.0-cp37-cp37m-manylinux2014_aarch64.whl (45.3 MB view hashes)
> * torch-1.8.0-cp37-cp37m-manylinux1_x86_64.whl (735.5 MB view hashes) 

In both versions, macOS, Windows, and the ARM-64 Linux builds are small -- so presumably lacking GPU support -- and the Linux-amd64 build is large. So I'm pretty sure the upgrading to the most recent will still have the same problem.

So what makes you think 1.11 solves the issue?

I know @dyt811 quoted these install instructions:

[1.8.1](https://web.archive.org/web/20210507170251/https://pytorch.org/get-started/locally/)

```
pip3 install torch==1.8.1+cpu torchvision==0.9.1+cpu torchaudio===0.8.1 -f https://download.pytorch.org/whl/torch_stable.html
```

<details>

![1 8 1](https://user-images.githubusercontent.com/987487/167489041-c568ff8f-e5d0-41ed-9d1c-cfbf89dbc838.png)

</details>

[1.10.0](https://web.archive.org/web/20211116200059/https://pytorch.org/get-started/locally/)


```
pip3 install torch torchvision torchaudio
```

<details>

![1 10](https://user-images.githubusercontent.com/987487/167489064-2d24d606-2297-46cb-a988-c1fd8058fd2a.png)

</details>

Which is I guess what imad why you think 1.10 should just solve the issue? But if I [look on pypi](https://pypi.org/project/torch/1.10.0/#files) I see the same large file:

> *  torch-1.11.0-cp37-cp37m-manylinux1_x86_64.whl (750.6 MB view hashes) 



But I'm not actually able to reproduce that, because Archive.org stripped out the JavaScript from that page, and on https://pytorch.org/get-started/previous-versions/ I see that the 1.10 instructions should be

```
pip install torch==1.10.1+cpu torchvision==0.11.2+cpu torchaudio==0.10.1 -f https://download.pytorch.org/whl/torch_stable.html
```


But I see on https://pytorch.org/get-started/locally/ that the current 1.11 linux+CPU instructions are actually

```
pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu
```

And it works great (notice: 169.2MB << 750MB):

```
(torch1) p115628@joplin:~/src$ pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cpu
Collecting torch
  Downloading https://download.pytorch.org/whl/cpu/torch-1.11.0%2Bcpu-cp38-cp38-linux_x86_64.whl (169.2 MB)
     |███████████████████▋            | 103.4 MB 87.8 MB/s eta 0:00:01^C
ERROR: Operation cancelled by user
```

Which **drops the `+cpu` syntax**, I guess enabled by switching to `-I` (`--extra-index-url`) over `-f` (`--find-links`).

It turns out this works **even for 1.8**:

```
(torch1) p115628@joplin:~/src$ pip3 install torch==1.8.0 torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cpu
Collecting torch==1.8.0
  Downloading https://download.pytorch.org/whl/cpu/torch-1.8.0%2Bcpu-cp38-cp38-linux_x86_64.whl (169.1 MB)
     |█████████████▌                  | 71.2 MB 885 kB/s eta 0:01:51^C
ERROR: Operation cancelled by user
```

So, I think in SCT we can be satisfied with switching to `--extra-index-url https://download.pytorch.org/whl/cpu`; we already [hide the previous hack](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/528dfe2f0c658f7a8a494ad2cc90c37c10c29b47/requirements.txt#L32-L41) so it's not user-visible anyway; switching that out for this lets us unpin and then our maintenance burden is gone.

tl;dr: torch's latest work has improved the situation, but you don't need to bump the version you're working against, all we in SCT have to do is add `--extra-index-url https://download.pytorch.org/whl/cpu` to our pip commands (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3790).

As for [your instructions](https://ivadomed.org/installation.html), it looks like they've given the same treatment to _all_ the versions they used to use "local version identifiers" (the `+` syntax) for; so you currently tell people to:

```
pip install --upgrade pip
pip install ivadomed
pip install torch==1.8.1+cu111 torchvision==0.9.1+cu111 --find-links https://download.pytorch.org/whl/torch_stable.html
```

You can cut that down to


```
pip install --upgrade pip
pip install --extra-index-url https://download.pytorch.org/whl/cu111 ivadomed
```

They haven't documented this super well yet, but it works.

kousu commented:
tl;dr: thanks to your keen eyes @dyt811 and @kanishk16, I was able to sidestep the issue we were having in SCT: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3790
